### PR TITLE
Py 312 pre-liminary support

### PR DESCRIPTION
**Pull request by jnoortheen**
## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

### CODE REVIEW

**Comment by jnoortheen** [comment]
If merged, lets keep the commits separate instead of squashing.

**Comment by jnoortheen** [comment]
reported here https://github.com/python/cpython/issues/105881

**Comment by jnoortheen** [comment]
It seems multiple slashes will no longer be resolved and be considered UNC paths from now on. So we will continue to skip the test from `py3.12` onwards

**Comment by jnoortheen** [comment]
@anki-code would you be able to test it?

**Comment by jnoortheen** [comment]
around 30m gets saved in tests with https://github.com/xonsh/xonsh/pull/5156/commits/3898d9e2e0ff3c368a7919316d31e56f290947d7
image

image

File: tests/test_parser.py [hunk 1]
```diff
@@ -168,6 +168,8 @@ def test_string_literal_concat(first_prefix, second_prefix, check_ast):
 
 
 def test_f_env_var(check_xonsh_ast):
+    if VER_MAJOR_MINOR > (3, 11):
+        pytest.xfail("f-string with special syntax are not supported yet")
```
Comment (gforsyth) [inline tests/test_parser.py:172]: Can we open a tracking issue for this so it doesn't get missed?

File: tests/xoreutils/test_which.py [hunk 2]
```diff
@@ -1,14 +1,15 @@
...
 from xonsh.tools import ON_WINDOWS
 from xonsh.xoreutils import _which
 
 
-class TestWhich:
+class TestWhich(TestCase):
     # Tests for the _whichgen function which is the only thing we
     # use from the _which.py module.
-    def setup(self):
+    def setUp(self):
```
Comment (gforsyth) [inline tests/xoreutils/test_which.py:12]:
```suggestion
    @classmethod
    def setup_class(cls):
```

I think this is the standard way to handle class-wrapped setup and teardown in pytest (will require swapping out all the `self` below, though).  Not a blocker -- I am not a fan of test classes anyway.

File: xonsh/tokenize.py [hunk 3]
```diff
@@ -83,9 +84,7 @@
 from xonsh.lazyasd import LazyObject
 from xonsh.platform import PYTHON_VERSION_INFO
 
-HAS_WALRUS = PYTHON_VERSION_INFO > (3, 8)
-if HAS_WALRUS:
-    from token import COLONEQUAL  # type:ignore
+HAS_WALRUS = True
```
Comment (gforsyth) [inline xonsh/tokenize.py:87]: I think we can remove this entirely, and then skip the checks on lines 128 and 235.

**Review by gforsyth — COMMENTED** [review]
This looks like a good start to me @jnoortheen!

I'd like to remove the remaining `HAS_WALRUS` checks, but except for that none of my other comments are blocking.

File: tests/test_parser.py (hunk 1 above)
Comment (jnoortheen) [reply tests/test_parser.py:172]: https://github.com/xonsh/xonsh/issues/5166 done

File: tests/xoreutils/test_which.py (hunk 2 above)
Comment (jnoortheen) [reply tests/xoreutils/test_which.py:12]: setup_class will run at class level only during the start of the test. It is a not straight equivalent to the setup of nosetest . For now we can use the unit-test and later refactor to a pytest if needed.

File: xonsh/tokenize.py (hunk 3 above)
Comment (jnoortheen) [reply xonsh/tokenize.py]: done

**Review by gforsyth — APPROVED** [review]
This looks good to me, @jnoortheen !

I'll leave it for you to merge in case there's anything you'd like to tweak before it goes in.

**Pull request merged**